akka-http-core/src/main/scala/akka/http/impl/engine/http2/client/PersistentConnection.scala

```diff
@@ -0,0 +1,127 @@
+import akka.stream.scaladsl.{ Flow, Source }
+import akka.stream.stage.{ GraphStage, GraphStageLogic, InHandler, OutHandler, StageLogging }
+import akka.stream.{ Attributes, FlowShape, Inlet, Outlet }
+
+/** INTERNAL API */
+@InternalApi
+private[http2] object PersistentConnection {
+  /**
+   * Wraps a connection flow with transparent reconnection support.
+   *
+   * Backpressure logic:
+   *
+   *  * when unconnected wait for pull on response side to pull request side
+   *  * when first request comes in, establish connection and send initial request to connection
+   *  * connection then drives backpressure logic, i.e.
+   *    * it may pull regardless of outer demand
+   *    * if it pulls and runs multiple requests concurrently, it must also buffer them internally
+   *    * this stage will only ever buffer requests directly in the incoming request slot
+   *    * outgoing responses are never buffer but we pull the connection only if there is outer demand
+   *
+   * Error reporting logic when connection breaks while requests are running:
+   *  * generate error responses with 502 status code
+   *  * custom attribute contains internal error information
+   */
+  def managedConnection(connectionFlow: Flow[HttpRequest, HttpResponse, Any]): Flow[HttpRequest, HttpResponse, NotUsed] =
+    Flow.fromGraph(new Stage(connectionFlow))
+
+  private class AssociationTag extends RequestResponseAssociation
+  private val associationTagKey = AttributeKey[AssociationTag]("PersistentConnection.associationTagKey")
+  private val errorResponse =
+    HttpResponse(
+      StatusCodes.BadGateway,
+      entity = "The server closed the connection before delivering a response.")
+
+  private class Stage(connectionFlow: Flow[HttpRequest, HttpResponse, Any]) extends GraphStage[FlowShape[HttpRequest, HttpResponse]] {
+    val requestIn = Inlet[HttpRequest]("PersistentConnection.requestIn")
+    val responseOut = Outlet[HttpResponse]("PersistentConnection.responseOut")
+
+    val shape: FlowShape[HttpRequest, HttpResponse] = FlowShape(requestIn, responseOut)
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with StageLogging {
+      become(Unconnected)
+
+      def become(state: State): Unit = setHandlers(requestIn, responseOut, state)
+
+      trait State extends InHandler with OutHandler
+      object Unconnected extends State {
+        override def onPush(): Unit = connect()
+        override def onPull(): Unit =
+          if (!hasBeenPulled(requestIn)) // requestIn might already have been pulled when we failed and went back to Unconnected
+            pull(requestIn)
+
+      }
+
+      def connect(): Unit = {
+        val requestOut = new SubSourceOutlet[HttpRequest]("PersistentConnection.requestOut")
+        val responseIn = new SubSinkInlet[HttpResponse]("PersistentConnection.responseIn")
+
+        Source.fromGraph(requestOut.source)
+          .via(connectionFlow)
```

Comment:
I guess re-try with different IP or such can be achieved with `Flow.lazy() => create actual flow)` but that may not be obvious so hint in docs about that would be good. 

Long term something for backoff would be good, but perhaps that can also be outsourced to some clever composition with the connection flow instead of added in here as well?

Reply:
Good point, might make sense to include that in this stage.

Reply:
To clarify, I think backoff would belong here. I wonder if migrating to different IPs should be part of the load-balancing part instead.